File: src/2025-02-24-deepclaude-rust/javascript/exercises/practice/beer-song/beer-song.js
//
// This is only a SKELETON file for the 'Beer Song' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const recite = (initialBottlesCount, takeDownCount) => {
  const verses = [];
  for (let i = 0; i < takeDownCount; i++) {
    const currentBottles = (initialBottlesCount - i + 100) % 100;
    const line1 = currentBottles === 0 
      ? "No more bottles of beer on the wall, no more bottles of beer."
      : `${currentBottles} ${currentBottles === 1 ? 'bottle' : 'bottles'} of beer on the wall, ${currentBottles} ${currentBottles === 1 ? 'bottle' : 'bottles'} of beer.`;

    let line2;
    if (currentBottles === 0) {
      line2 = "Go to the store and buy some more, 99 bottles of beer on the wall.";
    } else {
      const nextBottles = currentBottles - 1;
      const nextBottlesText = nextBottles === 0 
        ? 'no more bottles'
        : `${nextBottles} ${nextBottles === 1 ? 'bottle' : 'bottles'}`;
      line2 = `Take ${currentBottles === 1 ? 'it' : 'one'} down and pass it around, ${nextBottlesText} of beer on the wall.`;
    }

    verses.push(line1);
    verses.push(line2);
    if (i < takeDownCount - 1) {
      verses.push('');
    }
  }
  return verses;
};

File: src/2025-02-24-deepclaude-rust/go/exercises/practice/beer-song/beer_song.go
package beer

import (
	"errors"
	"fmt"
	"strings"
)

func Song() string {
	s, _ := Verses(99, 0)
	return s
}

func Verses(start, stop int) (string, error) {
	if start < stop {
		return "", errors.New("start must be greater than or equal to stop")
	}
	var verses []string
	for i := start; i >= stop; i-- {
		verse, err := Verse(i)
		if err != nil {
			return "", err
		}
		verses = append(verses, verse)
	}
	return strings.Join(verses, "\n\n") + "\n", nil
}

func Verse(n int) (string, error) {
	if n < 0 || n > 99 {
		return "", errors.New("invalid verse")
	}
	if n == 0 {
		return "No more bottles of beer on the wall, no more bottles of beer.\nGo to the store and buy some more, 99 bottles of beer on the wall.", nil
	}

	action := "Take one down and pass it around"
	if n == 1 {
		action = "Take it down and pass it around"
	}

	next := n - 1
	nextPhrase := nextBottlesPhrase(next)

	line1 := fmt.Sprintf("%s of beer on the wall, %s of beer.", bottlesPhrase(n), bottlesPhrase(n))
	line2 := fmt.Sprintf("%s, %s of beer on the wall.", action, nextPhrase)

	return strings.Join([]string{line1, line2}, "\n") + "\n", nil
}

func bottlesPhrase(n int) string {
	if n == 1 {
		return "1 bottle"
	}
	return fmt.Sprintf("%d bottles", n)
}

func nextBottlesPhrase(next int) string {
	if next == 0 {
		return "no more bottles"
	} else if next == 1 {
		return "1 bottle"
	}
	return fmt.Sprintf("%d bottles", next)
}
